dynamic shapes for combo_kenel/foreach_kernel

qchip opened the pull request:
This PR add dynamic shapes support to foreach and combo kernels for horizontal fusion.
A flag `combo_kernel_foreach_dynamic_shapes` (default False to avoid disturb production workflows) is added to _inductor/config.py. Setting it to True enables automatic dynamic shapes for foreach kernels. It is always enabled for combo kernels cases. Added unit cases.

This PR also fixes a flaky test case for [T198833257](https://www.internalfb.com/intern/tasks/?t=198833257) 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

qchip commented:
Hi @mlazos, any comments on the PR? we target to production it in the coming weeks, thanks for your help!

mlazos commented on test/inductor/test_foreach.py:
```diff
@@ -487,6 +487,43 @@ def fn(a0, a1, b0, b1):
+    @torch._dynamo.config.patch("assume_static_by_default", False)
+    @torch._inductor.config.patch("combo_kernel_foreach_dynamic_shapes", True)
+    def test_enable_dynamic_shapes_python_wrapper(self, op=torch._foreach_add):
+        def fn(a0, a1, b0, b1):
```
does this test actually use dynamic shapes?

qchip replied:
when assume_static_by_default==False and automatic_dynamic_shapes==True, the generated triton kernel code has xnumels/ynumels as the arguments in stead of hardcoding their values, so that it can work with input tensors of different shapes. This test itself runs with only one input tensor size (on the kernel that support variable sizes), the other test `test_dynamic_shapes_2d_blocking_round_robin` in test_combo_kernels.py actually tested it with two different sizes.

mlazos commented on test/inductor/test_foreach.py:
```diff
@@ -487,6 +487,43 @@ def fn(a0, a1, b0, b1):
+    @torch._dynamo.config.patch("assume_static_by_default", False)
+    @torch._inductor.config.patch("combo_kernel_foreach_dynamic_shapes", True)
+    @torch._inductor.config.patch("cpp_wrapper", True)
+    def test_enable_dynamic_shapes_cpp_wrapper_cuda(self, op=torch._foreach_add):
```
Same question as above

mlazos reviewed (approved):
Looks good! Just a couple of questions on the tests

qchip commented:
@pytorchbot merge

qchip committed:
dynamic shapes for combo_kenel/foreach_kernel

Pull Request resolved: https://github.com/pytorch/pytorch/pull/134287
ghstack-source-id: 237803326

Differential Revision: [D61168497](https://our.internmc.facebook.com/intern/diff/D61168497/)

qchip commented:
@pytorchbot merge
